### PR TITLE
add ESXi support with thin cloning support

### PR DESCRIPTION
**Pull request by moshloop**
This adds ESXi support back in, rather than do a datastore level copy, it uses vmdk-tools to thin clone the disks - in theory this could be extended to support linked clones as well.

### CODE REVIEW

**Comment by akutz** [comment]
/hold

I want to review this prior to it merging.

**Comment by akutz** [comment]
Hi @moshloop,

You don't need to manually invoke the e2e job. In fact, please don't. It ties up the CI infrastructure. The job is failing the initial test because the dependencies were not updated correctly. Please see the test log.

**Comment by moshloop** [comment]
Not sure why the E2E test failed, a control plane endpoint comes up, but with the wrong cert name:
`Get https://192.168.3.141:6443/api?timeout=32s: x509: certificate is valid for 10.96.0.1, 192.168.3.53, 192.168.3.53, 192.168.3.53, not 192.168.3.141`

/test pull-cluster-api-provider-vsphere-e2e

File: pkg/cloud/vsphere/context/cluster_context.go [hunk 1]
```diff
@@ -71,6 +71,7 @@ type ClusterContext struct {
 	machineClient         client.MachineInterface
 	user                  string
 	pass                  string
+	server                string
```
Comment (akutz) [inline pkg/cloud/vsphere/context/cluster_context.go]: Hi @moshloop,

Why add this when it's easily pulled from the ClusterConfig?

File: pkg/cloud/vsphere/context/cluster_context.go [hunk 2]
```diff
@@ -102,7 +103,7 @@ func NewClusterContext(params *ClusterContextParams) (*ClusterContext, error) {
 	if logr == nil {
 		logr = klogr.New().WithName("default-logger")
 	}
-	logr = logr.WithName(params.Cluster.APIVersion).WithName(params.Cluster.Namespace).WithName(params.Cluster.Name)
+	logr = logr.WithName(params.Cluster.Name)
```
Comment (akutz) [inline pkg/cloud/vsphere/context/cluster_context.go]: Hi @moshloop,

Per the conversation from #322, please leave this alone for now.

File: pkg/cloud/vsphere/context/cluster_context.go [hunk 3]
```diff
@@ -137,6 +138,7 @@ func NewClusterContext(params *ClusterContextParams) (*ClusterContext, error) {
 		machineClient:         machineClient,
 		user:                  user,
 		pass:                  pass,
+		server:                clusterConfig.VsphereServer,
```
Comment (akutz) [inline pkg/cloud/vsphere/context/cluster_context.go]: Hi @moshloop,

Please see my previous comment about the server.

File: pkg/cloud/vsphere/context/cluster_context.go [hunk 4]
```diff
@@ -175,6 +177,20 @@ func (c *ClusterContext) Pass() string {
 	return c.pass
 }
 
+// Server returns the vCenter or ESXi server address
+func (c *ClusterContext) Server() string {
+	return c.server
```
Comment (akutz) [inline pkg/cloud/vsphere/context/cluster_context.go]: Hi @moshloop,

Unnecessary.

File: pkg/cloud/vsphere/context/cluster_context.go [hunk 5]
```diff
@@ -175,6 +177,20 @@ func (c *ClusterContext) Pass() string {
...
+}
+
+func (c *ClusterContext) GetSSH() *SSH {
```
Comment (akutz) [inline pkg/cloud/vsphere/context/cluster_context.go]: Hi @moshloop,

I would like SSH to be its own package. Please refer to https://github.com/akutz/simple-k8s-test-env/tree/feature/sk8-app/pkg/net/ssh for an example of an SSH package that supports direct and bastion access.

File: pkg/cloud/vsphere/context/machine_context.go [hunk 6]
```diff
@@ -230,7 +230,7 @@ func (c *MachineContext) Patch() error {
 	c.Machine.Status.ProviderStatus = newStatus
 
 	if !reflect.DeepEqual(c.Machine.Status, c.MachineCopy.Status) {
-		c.Logger.V(1).Info("updating machine status")
+		c.Logger.V(2).Info("updating machine status", "task", c.MachineStatus.TaskRef)
```
Comment (akutz) [inline pkg/cloud/vsphere/context/machine_context.go]: Hi @moshloop,

I like including the task, but instead of as a field, let's optionally add it as a namespace to the logger. It makes more sense since the task may not be present at all.

File: pkg/cloud/vsphere/context/session.go [hunk 7]
```diff
@@ -45,7 +45,7 @@ func getOrCreateCachedSession(ctx *MachineContext) (*Session, error) {
 	datacenter := ctx.MachineConfig.MachineSpec.Datacenter
 	sessionKey := ctx.ClusterConfig.VsphereServer + ctx.User() + datacenter
 
-	if session, ok := sessionCache[sessionKey]; ok {
+	if session, ok := sessionCache[sessionKey]; ok && session.IsVC() {
```
Comment (akutz) [inline pkg/cloud/vsphere/context/session.go]: Hi @moshloop,

Great catch!

File: pkg/cloud/vsphere/context/ssh.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
+package context
```
Comment (akutz) [inline pkg/cloud/vsphere/context/ssh.go]: Hi @moshloop,

Please see my previous comment about SSH. Also, on new files please add the license header.

File: pkg/cloud/vsphere/services/govmomi/builder.go [hunk 9]
```diff
@@ -0,0 +1,463 @@
+package govmomi
```
Comment (akutz) [inline pkg/cloud/vsphere/services/govmomi/builder.go]: Hi @moshloop,

Please add the license header.

File: pkg/cloud/vsphere/services/govmomi/builder.go [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	"sigs.k8s.io/yaml"
+
+	"github.com/vmware/govmomi/find"
+
+	"context"
```
Comment (akutz) [inline pkg/cloud/vsphere/services/govmomi/builder.go]: Hi @moshloop,

Please organize imports as such:

```
import (
    STDLIB

    EXTERNAL

    KUBE+CAPI

    PROJECT
)
```

File: pkg/cloud/vsphere/services/govmomi/create.go [hunk 11]
```diff
@@ -256,34 +257,89 @@ func Create(ctx *context.MachineContext, bootstrapToken string) error {
...
+			return err
+		}
+		ctx.Logger.V(6).Info("vcenter cloning machine", "clone-spec", spec)
+		task, err = builder.Src.Clone(ctx, builder.Folder, ctx.Machine.Name, *spec)
+		if err != nil {
+			return errors.Wrapf(err, "error trigging clone op for machine %q", ctx)
+		}
+
+	} else {
+		// ESXi does not support cloning, we need to "ghetto" clone the VM
```
Comment (akutz) [inline pkg/cloud/vsphere/services/govmomi/create.go]: Hi @moshloop,

Please use a different word rather than "ghetto".

**Comment by figo** [comment]
> Not sure why the E2E test failed, a control plane endpoint comes up, but with the wrong cert name:
> `Get https://192.168.3.141:6443/api?timeout=32s: x509: certificate is valid for 10.96.0.1, 192.168.3.53, 192.168.3.53, 192.168.3.53, not 192.168.3.141`
> 
> /test pull-cluster-api-provider-vsphere-e2e
@moshloop 
haven't look at your code, but it may due to the host IP not been added to api endpoint cert, does your code affect it?

File: pkg/cloud/vsphere/services/govmomi/create.go (hunk 11 above)
Comment (moshloop) [reply pkg/cloud/vsphere/services/govmomi/create.go]: This is in reference to other ghetto tools e.g. https://github.com/lamw/ghettoVCB

**Comment by moshloop** [comment]
@figo 
> host IP not been added to api endpoint cert

Are the endpoints certs not generated *before* the host is created? i.e. we don't the ip(s) at cert creation time? How is this working on VMC?

In the other tooling I built, I used localhost and a "sidecar" based api-server loadbalancer for internal component to component traffic, 

For external -> load balancer, I added insecure or used token based auth,

File: pkg/cloud/vsphere/context/cluster_context.go (hunk 5 above)
Comment (moshloop) [reply pkg/cloud/vsphere/context/cluster_context.go]: I moved ssh into its own package, I don't think we *need* bastion access for ESXi, and the sk8 ssh package does not support username/password auth, it also has a few other deps that don't make it a simple copy and paste so I think it would better to leave this implementation as is, and bring on other features from sk8 as and when they are needed.

**Comment by moshloop** [comment]
@akutz I think this is ready to go

**Comment by akutz** [comment]
Hi @moshloop,

Reviewing...

File: pkg/cloud/vsphere/context/cluster_context.go [hunk 12]
```diff
@@ -186,6 +187,16 @@ func (c *ClusterContext) Pass() string {
 	return c.pass
 }
 
+// GetSSH returns a SSH client that can be used to execute commands against the vSphere endpoint
+func (c *ClusterContext) GetSSH() *ssh.SSH {
```
Comment (akutz) [inline pkg/cloud/vsphere/context/cluster_context.go]: Hi @moshloop,

What's the reason this needs to be part of the Cluster context? The `user` and `pass`? I think I'd rather defer this to a secret if possible.

File: pkg/cloud/vsphere/services/govmomi/esxi/clone.go [hunk 13]
```diff
@@ -17,12 +17,271 @@ limitations under the License.
...
 
 	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/cloud/vsphere/context"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/cloud/vsphere/services/govmomi/extra"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/cloud/vsphere/services/govmomi/template"
 )
 
 // Clone kicks off a clone operation on ESXi to create a new virtual machine.
 func Clone(ctx *context.MachineContext, userData []byte) error {
-	return errors.New("temporarily disabled esxi support")
+	vm, err := NewVmContext(ctx)
```
Comment (akutz) [inline pkg/cloud/vsphere/services/govmomi/esxi/clone.go]: Hi @moshloop,

Maybe this is `NewESXiLoggerContext`? Or is this sort of a combination of the VM builder you previously constructed?

File: pkg/cloud/vsphere/services/govmomi/esxi/clone.go [hunk 14]
```diff
@@ -17,12 +17,271 @@ limitations under the License.
...
+	if err != nil {
+		return errors.Wrapf(err, "Failed to create vm")
+	}
+	ctx.Logger.V(4).Info("Task submitted: " + task.Reference().Value)
+	ctx.MachineStatus.TaskRef = task.Reference().Value
+	return nil
+}
+
+// VmContext is a context that adds govmomi objects to a MachineContext
+type VmContext struct {
```
Comment (akutz) [inline pkg/cloud/vsphere/services/govmomi/esxi/clone.go]: Hi @moshloop,

Does this or any of its functions need to be exported?

File: pkg/cloud/vsphere/services/govmomi/esxi/clone.go [hunk 15]
```diff
@@ -17,12 +17,271 @@ limitations under the License.
...
+	Src        *object.VirtualMachine
+	Folder     *object.Folder
+	Datastore  *object.Datastore
+	Host       *object.HostSystem
+	Datacenter *object.Datacenter
+	Pool       *object.ResourcePool
+}
+
+// CreateVM submits the folder.createVM task
+func CreateVM(ctx *VmContext, spec *types.VirtualMachineConfigSpec) (*object.Task, error) {
```
Comment (akutz) [inline pkg/cloud/vsphere/services/govmomi/esxi/clone.go]: Hi @moshloop,

Does this need to be exported?

File: pkg/cloud/vsphere/services/govmomi/esxi/clone.go [hunk 16]
```diff
@@ -17,12 +17,271 @@ limitations under the License.
...
+
+	folders, err := ctx.Datacenter.Folders(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return folders.VmFolder.CreateVM(ctx, *spec, ctx.Pool, ctx.Host)
+}
+
+// NewVmContext creates a new VmContext from a MachineContext
+func NewVmContext(ctx *context.MachineContext) (*VmContext, error) {
```
Comment (akutz) [inline pkg/cloud/vsphere/services/govmomi/esxi/clone.go]: Hi @moshloop,

Does this need to be exported?

File: pkg/cloud/vsphere/services/govmomi/esxi/clone.go [hunk 17]
```diff
@@ -17,12 +17,271 @@ limitations under the License.
...
+func (ctx *VmContext) GetVirtualMachineMO() (*mo.VirtualMachine, error) {
+	var props mo.VirtualMachine
+	if err := ctx.Src.Properties(ctx.Context, ctx.Src.Reference(), nil, &props); err != nil {
+		return nil, err
+	}
+	return &props, nil
+}
+
+// GetNics creates and returns network devices from those specifid under Network.Devices
+func GetNics(ctx *VmContext) ([]types.BaseVirtualDeviceConfigSpec, error) {
```
Comment (akutz) [inline pkg/cloud/vsphere/services/govmomi/esxi/clone.go]: Hi @moshloop,

`s/GetNIcs/getNetworkSpecs` to be consistent with the vCenter package?

File: pkg/cloud/vsphere/services/govmomi/esxi/clone.go [hunk 18]
```diff
@@ -17,12 +17,271 @@ limitations under the License.
...
+		if err != nil {
+			return nil, err
+		}
+		devices = append(devices, netdev)
+	}
+	return devices.ConfigSpec(types.VirtualDeviceConfigSpecOperationAdd)
+}
+
+// GetSerial creates and returns a new serial port to help with entropy
+func GetSerialPort(ctx *VmContext, devices object.VirtualDeviceList) (object.VirtualDeviceList, error) {
```
Comment (akutz) [inline pkg/cloud/vsphere/services/govmomi/esxi/clone.go]: Hi @moshloop,

Please see my earlier questions about whether things need to be exported.

File: pkg/cloud/vsphere/services/govmomi/esxi/clone.go [hunk 19]
```diff
@@ -17,12 +17,271 @@ limitations under the License.
...
+	}
+	portSpec.Key = -302
+	portSpec.ControllerKey = c.Key
+	devices = append(devices, portSpec)
+
+	return devices, nil
+}
+
+// GetVirtualDisks returns the list of existing virtual disks on the source template
+func GetVirtualDisks(ctx *VmContext) ([]types.VirtualDisk, error) {
```
Comment (akutz) [inline pkg/cloud/vsphere/services/govmomi/esxi/clone.go]: Hi @moshloop,

Please see my earlier questions about whether things need to be exported.

Also, perhaps `s/GetVirtualDisks/getDiskSpec` to be consistent with the vCenter package? Or will there be more than one disk?

File: pkg/cloud/vsphere/services/metadata/constants.go [hunk 20]
```diff
@@ -16,7 +16,11 @@ limitations under the License.
 
 package metadata
 
-const format = `
+const basic = `
```
Comment (akutz) [inline pkg/cloud/vsphere/services/metadata/constants.go]: Hi @moshloop,

Why the change here? Whether it's ESXi or vCenter, the cloud-init bits should be the same on the inside of the guest, should they not?

File: pkg/cloud/vsphere/services/metadata/metadata.go [hunk 21]
```diff
@@ -29,6 +29,24 @@ import (
 // New returns the cloud-init metadata as a base-64 encoded string for a given
 // machine context.
 func New(ctx *context.MachineContext) ([]byte, error) {
+	net := ctx.MachineConfig.MachineSpec.Network
+
+	var format string
+
+	// skip network customization for basic DHCP networks reducing the risk of
```
Comment (akutz) [inline pkg/cloud/vsphere/services/metadata/metadata.go]: Hi @moshloop,

Hmm. I see. Good catch, but please yank this out of this PR and let's discuss it in a new PR. Please file an issue as well. This is a really good point -- I just don't want to tie it to the ESX work.

File: pkg/cloud/vsphere/services/ssh/ssh.go [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+	"fmt"
+	"io"
+	"os"
+
+	"golang.org/x/crypto/ssh"
+
+	"github.com/go-logr/logr"
+)
+
+type SSH struct {
```
Comment (akutz) [inline pkg/cloud/vsphere/services/ssh/ssh.go]: Hi @moshloop,

Please add comments to exported types/functions. Thanks!

File: pkg/cloud/vsphere/services/govmomi/esxi/clone.go (hunk 14 above)
Comment (moshloop) [reply pkg/cloud/vsphere/services/govmomi/esxi/clone.go]: So I renamed this to GovmomiContext and moved to a new context package, a different PR will refactor vcenter to also use this

File: pkg/cloud/vsphere/services/metadata/metadata.go (hunk 21 above)
Comment (moshloop) [reply pkg/cloud/vsphere/services/metadata/metadata.go]: So this is actually a dependency on getting ESXi to work, I don't think ESXi generates the mac address on create, but rather on power on, which is too late the metadata script.

**Comment by akutz** [comment]
Hi @moshloop,

Do you think you want to come back to this or should we close it?
